web/source/kmwkeyboards.ts

```diff
@@ -0,0 +1,1383 @@
+    if(options['context'] == 'keyboard') {
+      var i,kp=x['keyboard'];
+      // Process array of keyboard definitions
+      if(typeof(kp.length) == 'number') {
+        for(i=0; i<kp.length; i++) {
+          this.registerLanguagesForKeyboard(kp[i],options,i);
+        }
+      } else { // Process a single keyboard definition
+        this.registerLanguagesForKeyboard(kp,options,0);
+      }
+    } else if(options['context'] == 'language') { // Download the full list of supported keyboard languages
+      this.languageList = x['languages'];
+      if(this.languagesPending) {
+        this.addLanguageKeyboards(this.languagesPending);
+      }
+      this.languagesPending = [];      
+    }
+  }
+
+  /**
+   *  Add default or all keyboards for a given language
+   *  
+   *  @param  {Object}   languages    Array of language names
+   **/           
+  addLanguageKeyboards(languages) {
+    var i, j, lgName, cmd, first, addAll;
+    
+    // Defer registering keyboards by language until the language list has been loaded
+    if(this.languageList == null) {                       
+      first = (this.languagesPending.length == 0);
+
+      for(i=0; i<languages.length; i++) {
+        this.languagesPending.push(languages[i]);
+      }
+
+      if(first) {
+        this.keymanCloudRequest('',true);
+      }   
+    } else { // Identify and register each keyboard by language name
+      cmd = '';
+      for(i=0; i<languages.length; i++) {
+        lgName = languages[i].toLowerCase();
+        addAll = (lgName.substr(-1,1) == '$'); 
+        if(addAll) {
+          lgName = lgName.substr(0,lgName.length-1);
+        }
+
+        for(j=0; j<this.languageList.length; j++) {
+          if(lgName == this.languageList[j]['name'].toLowerCase()) {
+            if(cmd != '') {
+              cmd = cmd + ',';
+            }
+
+            cmd = cmd+'@'+this.languageList[j]['id'];
+            if(addAll) {
+              cmd = cmd + '$';
+            }
+
+            break;
+          }
+        }
+      }
+
+      if(cmd == '') {
+        this.keymanweb.util.alert('No keyboards are available for '+languages[0]+'. '
+          +'Does it have another language name?');
+      } else {
+        this.keymanCloudRequest('&keyboardid='+cmd,false);
+      }
+    }
+  }
+
+  /**
+   *  Request keyboard metadata from the Keyman Cloud keyboard metadata server
+   *   
+   *  @param  {string}   cmd        command string
+   *  @param  {boolean?} byLanguage if true, context=languages, else context=keyboards 
+   **/
+  keymanCloudRequest(cmd: string, byLanguage?: boolean) {         
+    var URL='https://r.keymanweb.com/api/4.0/', tFlag, 
+      Lscript = <HTMLScriptElement> this.keymanweb.util._CreateElement('SCRIPT');
+    
+    URL = URL + ((arguments.length > 1) && byLanguage ? 'languages' : 'keyboards')
+      +'?jsonp=keyman.register';  
+
+    var kbdManager = this;
+    
+    // Set callback timer
+    tFlag='&timerid='+window.setTimeout(
```

Comment:
This is really for a separate PR. Can I just say: "YUK". I know you didn't write this code. I also know why it exists. Because back in the bad old days, we couldn't always reliably detect if a script element loaded or failed, especially with Opera 9.x. Now that `script.onerror` and `script.onload` are guaranteed to fire after the script executes/fails to download, we should use those instead. Note: if a keyboard file loads but fails to parse, `onload` is still called, so the appropriate management here is to test that it registers and show UI if it fails to register by the time `script.onload` is called. If that all makes sense to you, we can create an issue to do this.

Reply:
Issue #535 is now up.